### PR TITLE
added initialization schemes in torch.nn.init

### LINKED ISSUE #101: weight initializations for conv2d and linear
Allow different weight initialization schemes for conv2d and linear.
See: https://github.com/Kaixhin/nninit for some schemes.

### PR DESCRIPTION
**Pull request by alykhantejani**
- Ported initialization schemed from https://github.com/alykhantejani/nninit to `torch.nn.init`
- Added the `TestNNInit` `TestCase` to `test_nn.py` 
- small PEP8 fixes (removed whitespace  etc.)

Once merged we can probably close #101

### CODE REVIEW

File: torch/nn/init.py [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+    Examples:
+        >>> w = torch.Tensor(3, 5)
+        >>> nninit.kaiming_normal(w, gain=np.sqrt(2.0))
+    """
+    if isinstance(tensor, Variable):
+        kaiming_normal(tensor.data, gain=gain)
+        return tensor
+    else:
+        fan_in, _ = _calculate_fan_in_and_fan_out(tensor)
+        std = gain * np.sqrt(1.0 / fan_in)
```
Comment (szagoruyko) [inline torch/nn/init.py]: there are 2 versions of this init, one divides by `nInputPlane*kw*kh`, another by `nOutputPlane*kw*kh`, see here https://github.com/facebook/fb.resnet.torch/issues/106
also, why didn't you make `np.sqrt(2.0)` default in this line?

Comment (alykhantejani) [inline torch/nn/init.py]: Hi @szagoruyko, 

That's interesting, I wasn't aware that there were two versions. What do you think is the best way to add this optionality to the API?

As for the comment " why didn't you make np.sqrt(2.0) default in this line?", I'm not sure I quite understand. Do you mean the default `gain` or that is should be `np.sqrt(2.0/fan_in)`?

Comment (szagoruyko) [inline torch/nn/init.py]: maybe a string argument for 'fan_in' or 'fan_out' defaulting to 'fan_in'?
I think it should be np.sqrt(2.0 / n) by default, but I see that all frameworks have different values there

**Comment by alykhantejani** [comment]
Darn, looks like the build fails because `scipy` isn't installed. I use these to do goodness of fit tests on the returned weights to a particular distribution. Do you know of  alternative package I can use for this, if we don't want to have scipy as a requirement?

**Comment by soumith** [comment]
@alykhantejani you can keep scipy parts in the tests. Do the following:

at the top of test_nn.py add these lines:

```
TEST_SCIPY = True
try:
    import scipy
except ImportError:
    TEST_SCIPY = False
```

Then, in the tests where you have scipy references, to the tests, add the annotation:

` @unittest.skipIf(not TEST_SCIPY, "SCIPY unavailable")`
Here's an example: https://github.com/pytorch/pytorch/blob/master/test/test_nn.py#L673

Then, here: https://github.com/pytorch/pytorch/blob/master/.travis.yml#L21
Add the line: `    - travis_retry pip install scipy`

File: torch/nn/init.py [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+    if isinstance(tensor, Variable):
+        xavier_normal(tensor.data, gain=gain)
+        return tensor
+    else:
+        fan_in, fan_out = _calculate_fan_in_and_fan_out(tensor)
+        std = gain * np.sqrt(2.0 / (fan_in + fan_out))
+        return tensor.normal_(0, std)
+
+
+def kaiming_uniform(tensor, gain=1):
```
Comment (colesbury) [inline torch/nn/init.py]: I don't see any mention of uniform initialization (other than Xavier) in referenced paper. Where is this described?

File: torch/nn/init.py (hunk 1 above)
Comment (colesbury) [inline torch/nn/init.py]: Yeah, it should be `sqrt(2.0 / n)` (the 2 instead of a 1 is the big takeaway from the paper). What's the point of gain? Isn't the point of the initialization that it preserves magnitudes? Setting a gain breaks that.

File: test/test_nn.py [hunk 3]
```diff
@@ -1450,6 +1458,233 @@ def test_batchnorm_eval(self):
...
+                    if as_variable:
+                        input_tensor = input_tensor.data
+
+                    tensor_shape = input_tensor.numpy().shape
+                    receptive_field = np.prod(tensor_shape[2:])
+
+                    expected_std = gain * np.sqrt(2.0 / ((tensor_shape[1] + tensor_shape[0]) * receptive_field))
+                    assert self._is_normal(input_tensor, 0, expected_std)
+
+    def test_kaiming_unifrom_errors_on_inputs_smaller_than_2d(self):
```
Comment (alykhantejani) [inline test/test_nn.py]: fix typo: unifrom -> uniform

File: torch/nn/init.py (hunk 2 above)
Comment (alykhantejani) [inline torch/nn/init.py]: It's not explicitly described in the paper, and in that particular work they use a Gaussian distribution. However, I believe the condition of the initialization scheme is that the weights be symmetrically distributed around zero, so a uniform distribution is also a valid choice with `std = gain * np.sqrt(1.0/n)`

File: torch/nn/init.py (hunk 1 above)
Comment (alykhantejani) [inline torch/nn/init.py]: I think the default gain should probably be set to `np.sqrt(2.0)`, as `1` isn't very useful.

`sqrt(2.0/n)` is valid for ReLU, but for leaky ReLU variants it should be `sqrt(2/n*(1 + a^2))`, which can be altered via the `gain` parameter. For example see discussions in section 2 of [Kaiming's paper](https://arxiv.org/pdf/1502.01852.pdf) for a discussion of PReLU

Comment (alykhantejani) [inline torch/nn/init.py]: I've updated the PR to make the default `gain = np.sqrt(2.0)` for kaiming initialization

**Comment by alykhantejani** [comment]
@soumith The `pip install scipy` in `.travis.yaml` seems to be failing for the 2.x builds. Any idea why?

**Comment by soumith** [comment]
hmmm, googling around seems to be a known issue.
Can you try:
https://github.com/DistrictDataLabs/yellowbrick/blob/master/.travis.yml#L7-L8

or:
https://github.com/travis-ci/travis-ci/issues/2650#issuecomment-266721196
https://github.com/travis-ci/travis-ci/issues/2650#issuecomment-252285413

File: torch/nn/init.py (hunk 1 above)
Comment (colesbury) [inline torch/nn/init.py]: I think it would make sense to take in `a` (initial coefficient for PReLU, where a=0 is the default). `gain` seems like the wrong level of abstraction. Of course if you put in the right "gain" it's equivalent (but so is just taking in the correct standard deviation).

File: torch/nn/init.py [hunk 4]
```diff
@@ -0,0 +1,240 @@
+import numpy as np
```
Comment (colesbury) [inline torch/nn/init.py]: numpy isn't currently a dependency of PyTorch. It looks like you can substitute `math` and tensor operations for the usages.

File: torch/nn/init.py [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+        fan_out = num_output_fmaps * receptive_field_size
+
+    return fan_in, fan_out
+
+
+def xavier_uniform(tensor, gain=1):
+    """Fills the input Tensor or Variable with values according to the method described in "Understanding the difficulty of training
+       deep feedforward neural networks" - Glorot, X. and Bengio, Y., using a uniform distribution.
+
+       The resulting tensor will have values sampled from U(-a, a) where a = gain * sqrt(2/(fan_in + fan_out))
```
Comment (colesbury) [inline torch/nn/init.py]: Is this missing a `sqrt(3)` in the definition of `a`?

**Review by colesbury — COMMENTED** [review]
This looks pretty good. I think the most important thing is to remove the numpy dependency, since PyTorch does not currently require numpy.

File: torch/nn/init.py (hunk 5 above)
Comment (alykhantejani) [inline torch/nn/init.py]: @colesbury good spot - will update this

File: torch/nn/init.py (hunk 1 above)
Comment (alykhantejani) [inline torch/nn/init.py]: @colesbury I'm also ok with this, I was trying to be consistent with the Lasagne format (http://lasagne.readthedocs.io/en/latest/modules/init.html). 

I guess we should then remove the `gain` input from the `xavier_*` functions as if you want a `gain != 1`  you probably want to use `kaiming_*`. Do you agree?

File: torch/nn/init.py (hunk 4 above)
Comment (alykhantejani) [inline torch/nn/init.py]: @colesbury ah ok, sure I will make these changes too - thanks!

File: torch/nn/init.py [hunk 6]
```diff
@@ -69,7 +73,7 @@ def _calculate_fan_in_and_fan_out(tensor):
     else:
         num_input_fmaps = tensor.size(1)
         num_output_fmaps = tensor.size(0)
-        receptive_field_size = np.prod(tensor.numpy().shape[2:])
+        receptive_field_size = reduce(mul, (tensor.numpy().shape[2:]))
```
Comment (fmassa) [inline torch/nn/init.py]: Can we remove the `.numpy()` call?
Also, we could probably remove the need of the `reduce` by selecting the first two dimensions and then using `.view(-1).size(0)` instead.

File: torch/nn/init.py [hunk 7]
```diff
@@ -195,14 +199,14 @@ def orthogonal(tensor, gain=1):
         if tensor.ndimension() < 2:
             raise ValueError("Only tensors with 2 or more dimensions are supported.")
 
-        flattened_shape = (tensor.size(0), int(np.prod(tensor.numpy().shape[1:])))
+        flattened_shape = (tensor.size(0), int(reduce(mul, tensor.numpy().shape[1:])))
```
Comment (fmassa) [inline torch/nn/init.py]: Mabe remove `.numpy()`?

File: test/test_nn.py [hunk 8]
```diff
@@ -1689,13 +1693,13 @@ def test_orthogonal(self):
                         input_tensor = input_tensor.data
 
                     rows, cols = tensor_size[0], int(np.prod(tensor_size[1:]))
-                    flattened_tensor = input_tensor.view(rows, cols).numpy()
+                    flattened_tensor = input_tensor.view(rows, cols)
                     if rows > cols:
-                        assert np.allclose(np.dot(flattened_tensor.T, flattened_tensor), np.eye(cols) * gain ** 2,
-                                           atol=1e-6)
+                        assert torch.dist(torch.mm(flattened_tensor.t(), flattened_tensor),
```
Comment (fmassa) [inline test/test_nn.py]: I think you can pass a tolerance parameter to `self.assertEqual` that handles tensors as well, see https://github.com/pytorch/pytorch/blob/master/test/common.py#L93

File: torch/nn/init.py (hunk 6 above)
Comment (alykhantejani) [inline torch/nn/init.py]: Ah yes, good idea - I'll change that. Thanks

File: test/test_nn.py (hunk 8 above)
Comment (alykhantejani) [inline test/test_nn.py]: Great, I'll make that change. Thanks!

File: torch/nn/init.py [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+    if tensor.ndimension() < 2:
+        raise ValueError("fan in and fan out can not be computed for tensor of size ", tensor.size())
+
+    if tensor.ndimension() == 2:  # Linear
+        fan_in = tensor.size(1)
+        fan_out = tensor.size(0)
+    else:
+        num_input_fmaps = tensor.size(1)
+        num_output_fmaps = tensor.size(0)
+        receptive_field_size = reduce(mul, (tensor.numpy().shape[2:]))
```
Comment (apaszke) [inline torch/nn/init.py]: `reduce(mul, tensor.size()[2:])`

File: torch/nn/init.py [hunk 10]
```diff
@@ -0,0 +1,245 @@
...
+        b: the upper bound of the uniform distribution
+
+    Examples:
+        >>> w = torch.Tensor(3, 5)
+        >>> nninit.uniform(w)
+    """
+    if isinstance(tensor, Variable):
+        uniform(tensor.data, a=a, b=b)
+        return tensor
+    else:
```
Comment (apaszke) [inline torch/nn/init.py]: nit: you could remove the `else:` and move the actual implementation one indentation level back. I think that'd look nicer.

File: torch/nn/init.py (hunk 9 above)
Comment (alykhantejani) [inline torch/nn/init.py]: done, actually removed this by using `view(-1)` and `size(0)`

Comment (fmassa) [inline torch/nn/init.py]: Cool, thanks @alykhantejani ! On a side note, I just realized that this is actually equivalent to `.numel()`, but I think this is good as is.

File: torch/nn/init.py (hunk 7 above)
Comment (alykhantejani) [inline torch/nn/init.py]: done

File: test/test_nn.py (hunk 8 above)
Comment (alykhantejani) [inline test/test_nn.py]: done

File: torch/nn/init.py (hunk 10 above)
Comment (alykhantejani) [inline torch/nn/init.py]: I agree. Have changed this.

File: torch/nn/init.py (hunk 6 above)
Comment (alykhantejani) [inline torch/nn/init.py]: done

File: torch/nn/init.py (hunk 5 above)
Comment (alykhantejani) [inline torch/nn/init.py]: done

File: torch/nn/init.py (hunk 9 above)
Comment (apaszke) [inline torch/nn/init.py]: Yeah numel would be better I think, but it's ok

File: torch/nn/init.py (hunk 4 above)
Comment (alykhantejani) [inline torch/nn/init.py]: Have now removed the dependencies on `numpy`

File: torch/nn/init.py (hunk 9 above)
Comment (alykhantejani) [inline torch/nn/init.py]: Actually, `numel` reads more clearly. I'll make this change

**Comment by alykhantejani** [comment]
I've answered most reviews, but I still think there are the following open questions:

1. As @szagoruyko pointed out, in Kaiming initialization, there are two variants where one divides by `nInputPlane*kw*kh`, another by `nOutputPlane*kw*kh` ([see here](https://github.com/facebook/fb.resnet.torch/issues/106)). The method in this PR divides by `nInputPlane*kw*kh` (fan_in). Do we want to support both? If so, what's the best way to modify the interface i.e. should we take an additional param `use_fan_out=False`?

2. Should we remove the `gain` parameter from the `xavier_*` functions? I initially added it as this is what the Lasagne implementation has (used as reference). However, I think for cases where `gain != 1`, you would use `kaiming_*` instead?

3. Is it still the preferred method to rebase and squash all commits? If so, I will do this once we are all happy (so no more commits go in)

**Comment by apaszke** [comment]
1. I think adding a `use_fan_out` flag makes sense. But if Kaiming says that he used nOutputPlane, I think we should use that as the default.
2. No strong opinions on that. I can't see the value, but I don't have a problem either. It's already implemented so I think we can leave it.
3. No need to squash yourself, we can do that on GitHub while merging the PR.

File: test/test_nn.py [hunk 11]
```diff
@@ -1450,6 +1457,261 @@ def test_batchnorm_eval(self):
...
+                    else:
+                        a = 0
+                        init.kaiming_uniform(input_tensor)
+
+                    if as_variable:
+                        input_tensor = input_tensor.data
+
+                    fan_in = input_tensor.size(1)
+                    if input_tensor.dim() > 2:
+                        fan_in *= input_tensor[0][0].numel()
```
Comment (apaszke) [inline test/test_nn.py]: nit: you can do that index in one step: `input_tensor[0, 0]`. You don't have to fix that right now, but keep that in mind.

**Review by apaszke — APPROVED** [review]

**Comment by szagoruyko** [comment]
regarding fan_out, fan_in stabilizes forward activations, fan_out stabilizes gradients on backward, so it depends on architecture which one works better.
I think we should have a string argument so that we can extend it in future, eg. tf also has `fan_avg = (fan_in + fan_out) / 2`

**Comment by apaszke** [comment]
String arguments sound good to me.

File: test/test_nn.py (hunk 11 above)
Comment (alykhantejani) [inline test/test_nn.py]: updated these as I was making changes anyway

**Comment by alykhantejani** [comment]
@szagoruyko @apaszke I've now added a `mode` kwarg to the `kaiming_*` initializers.

**Comment by apaszke** [comment]
Thank you!
